### PR TITLE
Add implementation and tests for the sql REVERSE function.

### PR DESCRIPTION
**Pull request by Honeypot95**
Implementation of the sql REVERSE function. Code review is greatly appreciated. Also, REVERSE was listed twice in the main issue about what is missing.

### CODE REVIEW

File: src/function/scalar/string/reverse.cpp [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+	size_t n = strlen(input); // the size of the string in bytes
+	size_t bytes = 0;
+
+	output[n] = 0;
+
+	while (*input) {
+		if (!(*input & 0x80)) { // !*input & 0b10000000
+			bytes = 1;
+		} else if ((*input & 0xe0) == 0xc0) { // (*input & 0b1110_0000 == 0b1100_0000)
```
Comment (hannesmuehleisen) [inline src/function/scalar/string/reverse.cpp]: This is trying to compute UTF char length, right? There is code in value.cpp (IsUTF8String) that does something similar, perhaps those functions can be either merged or the UTFCharLen function should be moved there.

File: test/sql/function/test_stringfunctions.cpp [hunk 2]
```diff
@@ -131,3 +131,20 @@ TEST_CASE("UPPER/LOWER test", "[function]") {
...
+
+	REQUIRE_NO_FAIL(con.Query("CREATE TABLE strings(a STRING, b STRING)"));
+	REQUIRE_NO_FAIL(con.Query("INSERT INTO strings VALUES ('Hello', 'World'), "
+	                          "('HuLlD', NULL), ('MotörHead','RÄcks')"));
+
+	result = con.Query("select REVERSE(a) FROM strings");
+	REQUIRE(CHECK_COLUMN(result, 0, {"olleH", "DlLuH", "daeHrötoM"}));
+
+	result = con.Query("select REVERSE(b) FROM strings");
+	REQUIRE(CHECK_COLUMN(result, 0, {"dlroW", Value(), "skcÄR"}));
```
Comment (hannesmuehleisen) [inline test/sql/function/test_stringfunctions.cpp]: Perhaps also check with invalid UTF strings?

**Comment by hannesmuehleisen** [comment]
Thanks for your PR!

File: src/function/scalar/string/reverse.cpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+#include <string.h>
+
+using namespace std;
+
+namespace duckdb {
+
+typedef void (*str_function)(const char *input, char *output);
+
+static void strreverse(const char *input, char *output) {
+
```
Comment (hannesmuehleisen) [inline src/function/scalar/string/reverse.cpp]: Please also remove this blank line, CodeFactor complained

File: test/sql/function/test_stringfunctions.cpp (hunk 2 above)
Comment (hannesmuehleisen) [reply test/sql/function/test_stringfunctions.cpp]: Also missing the empty string as a test.

**Review by hannes — CHANGES_REQUESTED** [review]

File: test/sql/function/test_stringfunctions.cpp (hunk 2 above)
Comment (Mytherin) [reply test/sql/function/test_stringfunctions.cpp]: Invalid UTF strings should not be inside the database, so they should not be checked here. Empty string is a good suggestion though!

File: src/function/scalar/string/reverse.cpp (hunk 1 above)
Comment (Honeypot95) [reply src/function/scalar/string/reverse.cpp]: Yes, it appears very similar.
My proposed back-of-the-envelope solution is to have two additional functions:
UTF8charlen(char c); that returns 1 2 3 or 4 and -1 for error, and UTF8isvalid(char* c); that returns true or false if the present memory position is a valid utf8 string, using UTF8charlen inside.

Comment (Honeypot95) [reply src/function/scalar/string/reverse.cpp]: is a valid UTF8 code point, not string. The string check should still be done by the IsUTF8String function, using the helper functions.

Comment (Honeypot95) [reply src/function/scalar/string/reverse.cpp]: This is the solution I came up with:

```
/**
 * returns:
 *  0 for valid, [1, 4] for valid and a length, -1 for invalid code point
 */
int Value::UTF8CodeLength(const char *s) {
	int c = *s;

	if (s == nullptr) {
		return 0;
	}
	if (*s == '\200' && s[1] == '\0') {
		return 0; /* str_nil */
	}

	if ((c & 0x80) == 0)
		return 1; 	// valid with length 1
	if ((*s++ & 0xC0) != 0x80)
		return -1;
	if ((c & 0xE0) == 0xC0)
		return 2;	// valid with length 2
	if ((*s++ & 0xC0) != 0x80)
		return -1;
	if ((c & 0xF0) == 0xE0)
		return 3;	// valid with length 3
	if ((*s++ & 0xC0) != 0x80)
		return -1;
	if ((c & 0xF8) == 0xF0)
		return 4;	// valid with length 4
	return -1;
}

// adapted from MonetDB's str.c
bool Value::IsUTF8String(const char *s) {
	int c;
	int len;

	len = UTF8CodeLength(s);

	if (len == 0) {
		return true;
	}
	
	while ((len = UTF8CodeLength(s)) > 0) {
		s += len;
	}
	return len > -1 ? true : false;
}
```

The AddressSanitizer complains about the code though, so I cannot submit it.

**Pull request closed without merging**